### PR TITLE
feat: cache asm secrets

### LINKED ISSUE #1782: Add caching SecretsResolver/SecretsProvider
In production we use ASM to provide secrets, and we retrieve directly from it. This is costly and slow, especially for listing, so we should add a caching layer to the controller. The tricky bit is that we don't want to cache these results in the database because security, but with replicated controllers we'll need some solution to avoid every controller polling constantly.

One option might be for a single controller to perform the cache refreshing from ASM (holding a lease), and all other controllers redirect requests to that one for cache refreshes. We might need to extend leases to include an `owner` field (or more generically, `metadata`), that would contain the leader endpoint.

### PR DESCRIPTION
**Pull request by matt2e**
closes #1782
ASM changes:
- `ASM` uses `leader` package to coordinate between leader and followers
- The leader syncs secrets from ASM
    - Only requests secrets that have been updated since last sync
- Followers always uses admin service to get the secrets from the leader
- Both leader and follower cache secrets using `secretsCache`

Other changes:
- Remove additional db connection pools created at start up. Now a single db pool is created along with a dal at start up, which is then passed into the controller service and anything else that needs it

### CODE REVIEW

**Comment by alecthomas** [comment]
Wasn't this going to be two separate PRs?

**Comment by alecthomas** [comment]
Can you split it up please?

**Comment by matt2e** [comment]
I'll add some more tests to this PR too

File: common/configuration/asm_follower.go [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+	resp, err := f.client.SecretsList(ctx, connect.NewRequest(&ftlv1.ListSecretsRequest{
+		Module:        &module,
+		IncludeValues: &includeValues,
+	}))
+	if err != nil {
+		return nil, err
+	}
+	entries := []Entry{}
+	for _, s := range resp.Msg.Secrets {
+		components := strings.Split(s.RefPath, ".")
```
Comment (gak) [inline common/configuration/asm_follower.go:34]: I've seen this pattern a few times, maybe a `fn (rp *) ToRef (Ref, error) {}` ?

Comment (matt2e) [reply common/configuration/asm_follower.go:34]: I'll make a helper func for this 👍

Comment (matt2e) [reply common/configuration/asm_follower.go:34]: `ParseRef` already exists 🤦‍♂️

**Comment by matt2e** [comment]
> I'll add some more tests to this PR too

added integration test for syncing secrets

**Review by alecthomas — APPROVED** [review]

**Review by gak — APPROVED** [review]

File: common/configuration/asm_leader.go [hunk 2]
```diff
@@ -0,0 +1,315 @@
...
+	ref Ref
+	// value is nil when the secret was deleted
+	value optional.Option[[]byte]
+}
+
+type asmLeader struct {
+	client *secretsmanager.Client
+
+	// indicates if the initial sync with ASM has finished
+	loaded atomic.Value[bool]
```
[the remark this answers is not in the document]
Comment (matt2e) [reply common/configuration/asm_leader.go]: Oh nice

File: common/configuration/asm_leader.go [hunk 3]
```diff
@@ -0,0 +1,315 @@
...
+
+		case e := <-events:
+			l.processEvent(e)
+
+		case <-clock.After(clock.Until(nextSync)):
+			nextSync = clock.Now().Add(syncInterval)
+
+			err := l.sync(ctx)
+			if err == nil {
+				backOff = syncInitialBackoff
```
[the remark this answers is not in the document]
Comment (matt2e) [reply common/configuration/asm_leader.go:102]: This case is for when there is no error and is resetting the backoff. The error is logged below

**Comment by matt2e** [comment]
Updated: 
- added `secretsCache` which `asmLeader` and `asmFollower` use
- made the integration test where set secrets via ftl and ASM directly handle both cases:
    - setting/getting via leader which calls ASM
    - setting/getting via a follower, which calls a leader, which calls ASM

**Pull request merged**